File: 02-linkedlist/0204-24.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        h = head
        while h and h.next:
            h.val, h.next.val = h.next.val, h.val
            h = h.next.next
        return head
        

File: 02-linkedlist/0204-24.cpp
/*
 * leetcode 24
 * swap nodes in pairs
 * 
 */

class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        ListNode* dummyHead = new ListNode(0);
        dummyHead->next = head;
        ListNode* cur = dummyHead;
        while(cur->next != nullptr && cur->next->next != nullptr) {
            ListNode* tmp = cur->next; // 1
            ListNode* tmp2 = cur->next->next->next; // 3

            cur->next = cur->next->next; // 2
            cur->next->next = tmp; // reverse pair 1
            cur->next->next->next = tmp2; // connect to 3 (the start of next pair)
            // prepare for the next move
            cur = cur->next->next;
        }
        return dummyHead->next;        
    }
};

